lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -2610,3 +2611,457 @@ SILFunction *swift::lookupPrespecializedSymbol(SILModule &M,
+CanSILFunctionType ExistentialSpecializerTransform::createExistentialSpecializedFunctionType() {
+
+  CanSILFunctionType FTy = F->getLoweredFunctionType();
+  auto ExpectedFTy = F->getLoweredType().castTo<SILFunctionType>();
+  auto *Mod = F->getModule().getSwiftModule();
+  auto &Ctx = Mod->getASTContext();
+  GenericSignature *NewGenericSig;
+  GenericEnvironment *NewGenericEnv;
+
+  // Form a new generic signature based on the old one.
+  GenericSignatureBuilder Builder(Ctx);
+
+  /// If the original function is generic, then maintain the same.
+  /// Does it have generic structure already ?
+  auto HasGenericSignature = FTy->getGenericSignature() != nullptr;
+  if (HasGenericSignature) {
+    auto OrigGenericSig = FTy->getGenericSignature();
+    /// First, add the old generic signature.
+    Builder.addGenericSignature(OrigGenericSig);
+  }
+
+  /// Original list of parameters
+  SmallVector<SILParameterInfo, 4> params;
+  params.append(ExpectedFTy->getParameters().begin(),
+                ExpectedFTy->getParameters().end());
+
+  int GPIdx = 0;
+  /// Convert the protocol arguments of F to generic ones.
+  for (auto const& IdxIt: ExistentialArgDescriptor) { 
+    int Idx = IdxIt.first;
+    auto &param = params[Idx];
+    auto PType = param.getType();
+
+    assert(PType->is<ProtocolType>() ||
+      PType->is<ProtocolCompositionType>() /*|| PType->is<ClassType>()*/);
+    /// Generate new generic parameter.
+    auto *NewGenericParam = GenericTypeParamType::get(0, GPIdx++, Ctx);
+    Builder.addGenericParameter(NewGenericParam);
+    SmallVector<Requirement, 4> Requirements;
+    if (PType->is<ProtocolType>()) { 
+      auto DeclType = PType->castTo<ProtocolType>()->getDecl();
+      Requirement NewRequirement(RequirementKind::Conformance, NewGenericParam, DeclType->getDeclaredType());
+      Requirements.push_back(NewRequirement);
+    } else { 
+      auto CompType = PType->castTo<ProtocolCompositionType>();
+      recursivelyAddCompositeTypes(CompType, NewGenericParam, Requirements);
+    }
+    auto Source = GenericSignatureBuilder::FloatingRequirementSource::forAbstract();
+    for (auto &Req : Requirements) {
+      Builder.addRequirement(Req, Source, Mod);
+    }
+    Arg2GenericTypeMap[Idx] = NewGenericParam;
+    Arg2RequirementsMap[Idx] = Requirements;
+  }
+
+  /// Compute the generic signature.
+  NewGenericSig = std::move(Builder).computeGenericSignature(SourceLoc(), true);
+  NewGenericEnv = NewGenericSig->createGenericEnvironment();
+
+  /// Create a lambda for GenericParams.
+  auto getCanonicalType = [&](Type t) -> CanType {
+      return t->getCanonicalType(NewGenericSig);
+  };
+
+  /// Create the complete list of parameters.
+  int Idx = 0;
+  llvm::SmallVector<SILParameterInfo, 8> InterfaceParams;
+  InterfaceParams.reserve(params.size());
+  for (auto &param : params) {
+    if (Arg2GenericTypeMap[Idx]) {
+      auto GenericParam = Arg2GenericTypeMap[Idx];
```

Comment:
You should just store CanTypes in this DenseMap then you won't need to call getCanonicalType()

Reply:
There are a number os usage scenario. For now, I am keeping it as is. Let me know if you see a major issue.